File: solver.c
// Author: Armit 
// UpdateTime: 2022/09/24 

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#include <time.h>

// solver target
int m = 0;       // 1~12
int d = 0;       // 1~31
int w = 0;       // 1~7

// debug or perfcount
char* bin;
bool debug = false;   // whether show genegrated piece variations
int B_WIDTH = 8;      // block width in display
time_t ts_start;      // ts from DFS() start
time_t ts_last;       // ts from last solution found

// readable name mapping
char* MN[] = { "", 
  "一月", "二月", "三月", "四月", "五月", "六月",
  "七月", "八月", "九月", "十月", "十一月", "十二月",
};
char* DN[] = { "", 
  "1日", "2日", "3日", "4日", "5日", "6日", "7日", "8日", "9日", "10日",
  "11日", "12日", "13日", "14日", "15日", "16日", "17日", "18日", "19日", "20日",
  "21日", "22日", "23日", "24日", "25日", "26日", "27日", "28日", "29日", "30日",
  "31日",
};
char* WN[] = { "", 
  "星期一", "星期二", "星期三", "星期四", "星期五", "星期六", "星期日",
};

// board definition
const int nrow = 8;
const int ncol = 7;
typedef int board_t[8][7];      // C even NOT support const-var as array-length :(
enum BoardType { M, D, W, _ };
board_t board_type = {
  { M, M, M, M, M, M, _ },
  { M, M, M, M, M, M, _ },
  { D, D, D, D, D, D, D },
  { D, D, D, D, D, D, D },
  { D, D, D, D, D, D, D },
  { D, D, D, D, D, D, D },
  { D, D, D, W, W, W, W },
  { _, _, _, _, W, W, W },
};
board_t board_value = {
  {  1,  2,  3,  4,  5,  6,  0 },
  {  7,  8,  9, 10, 11, 12,  0 },
  {  1,  2,  3,  4,  5,  6,  7 },
  {  8,  9, 10, 11, 12, 13, 14 },
  { 15, 16, 17, 18, 19, 20, 21 },
  { 22, 23, 24, 25, 26, 27, 28 },
  { 29, 30, 31,  7,  1,  2,  3 },
  {  0,  0,  0,  0,  4,  5,  6 },
};
board_t mask = { 0 };           // aka. current partial solution; value v indicates the v-th piece is placed (index from 1), 0 for blank, -1 for invalid
board_t target = { 0 };         // 1 for masked, 0 for blank, -1 for invalid

// piece definition
const int npieces = 10;
typedef struct _piece_t {
  int row, col;
  bool* block;                  // store as 1d-array
  struct _piece_t *next;
} piece_t;
bool p0[] = {
  1, 1, 1, 1
};
bool p1[] = {
  1, 1, 1, 1,
  1, 0, 0, 0,
};
bool p2[] = {
  0, 1, 1, 1,
  1, 1, 0, 0,
};
bool p3[] = {
  1, 1, 1,
  1, 0, 0,
};
bool p4[] = {
  0, 1, 1,
  1, 1, 0,
};
bool p5[] = {
  1, 1, 1,
  1, 0, 1,
};
bool p6[] = {
  1, 1, 1,
  1, 1, 0,
};
bool p7[] = {
  1, 1, 1,
  1, 0, 0,
  1, 0, 0,
};
bool p8[] = {
  1, 1, 1,
  0, 1, 0,
  0, 1, 0,
};
bool p9[] = {
  0, 1, 1,
  0, 1, 0,
  1, 1, 0,
};
piece_t proto[] = {             // piece prototypes
  {1, 4, p0, NULL},
  {2, 4, p1, NULL},
  {2, 4, p2, NULL},
  {2, 3, p3, NULL},
  {2, 3, p4, NULL},
  {2, 3, p5, NULL},
  {2, 3, p6, NULL},
  {3, 3, p7, NULL},
  {3, 3, p8, NULL},
  {3, 3, p9, NULL},
};
piece_t* pieces[10] = { NULL };   // piece variations (saved in a linked-forward-star data structure)

// util functions
static inline bool p_get(piece_t *p, int x, int y) {
  return p->block[p->col * x + y];
}
static inline void p_set(piece_t *p, int x, int y, bool v) {
  p->block[p->col * x + y] = v;
}
static inline piece_t* p_copy(piece_t* p) {
  piece_t* r = malloc(sizeof(piece_t));
  r->row = p->row;
  r->col = p->col;
  r->next = NULL;
  size_t sz = sizeof(bool) * r->row * r->col;
  r->block = malloc(sz);
  memcpy(r->block, p->block, sz);
  return r;
}
static inline piece_t* p_rotate(piece_t *p) {
  piece_t* r = malloc(sizeof(piece_t));
  r->row = p->col;
  r->col = p->row;
  r->next = NULL;
  r->block = malloc(sizeof(bool) * r->row * r->col);
  for (int i = 0; i < r->row; i++)
    for (int j = 0; j < r->col; j++)
      p_set(r, i, j, p_get(p, j, r->row - i - 1));
  free(p);
  return r;
}
static inline piece_t* p_mirror(piece_t *p) {
  piece_t* r = malloc(sizeof(piece_t));
  r->row = p->row;
  r->col = p->col;
  r->next = NULL;
  r->block = malloc(sizeof(bool) * r->row * r->col);
  for (int i = 0; i < r->row; i++)
    for (int j = 0; j < r->col; j++)
      p_set(r, i, j, p_get(p, i, r->col - j - 1));
  free(p);
  return r;
}
static inline bool p_eq(piece_t *p1, piece_t *p2) {
  if (p1->row != p2->row) return false;
  if (p1->col != p2->col) return false;
  for (int i = 0; i < p1->row; i++)
    for (int j = 0; j < p1->col; j++)
      if (p_get(p1, i, j) != p_get(p2, i, j))
        return false;
  return true;
}
static inline void p_collect(piece_t *p, int k) {
  // check in list
  piece_t* q = pieces[k];
  while (q) {
    if (p_eq(q, p)) return;
    q = q->next;
  }

  // prepend
  q = p_copy(p);
  q->next = pieces[k];
  pieces[k] = q;
}
static inline bool p_can_put(piece_t *p, int x, int y) {
  for (int i = 0; i < p->row; i++)
    for (int j = 0; j < p->col; j++)
      if (p_get(p, i, j) && (mask[x + i][y + j] || !target[x + i][y + j]))
        return false;
  return true;
}
static inline void p_put(piece_t *p, int x, int y, int v) {
  for (int i = 0; i < p->row; i++)
    for (int j = 0; j < p->col; j++)
      if (p_get(p, i, j))
        mask[x + i][y + j] = v;     // set value=v
}
static inline void p_unput(piece_t *p, int x, int y) {
  for (int i = 0; i < p->row; i++)
    for (int j = 0; j < p->col; j++)
      if (p_get(p, i, j))
        mask[x + i][y + j] = 0;     // clear value
}

static inline int b_width(int x, int y) {
  int w = 0;
  int v = board_value[x][y];
  switch (board_type[x][y]) {
    case M: w = _mbstrlen(MN[v]) / 3 * 2; break;
    case D: w = _mbstrlen(DN[v]) - 1    ; break;
    case W: w = _mbstrlen(WN[v]) / 3 * 2; break;
  }
  return w;
}
static inline void print_board() {
  puts("[Board]");
  for (int i = 0; i < nrow; i++) {
    putchar('|');
    for (int j = 0; j < ncol; j++) {
      int v = board_value[i][j];
      char *s = MN[0];
      switch (board_type[i][j]) {
        case M: s = MN[v]; break;
        case D: s = DN[v]; break;
        case W: s = WN[v]; break;
      }
      int w = b_width(i, j), d = B_WIDTH - w;
      int pr = d / 2, pl = d - pr;
      while (pl-- > 0) putchar(' ');
      printf(s);
      while (pr-- > 0) putchar(' ');
      putchar('|');
    }
    putchar('\n');
  }
}
static inline void print_solution() {
  puts("[Solution]");
  for (int i = 0; i < nrow; i++) {
    putchar('|');
    for (int j = 0; j < ncol; j++)
      if (mask[i][j] > 0)
        printf(" %d |", mask[i][j] - 1);
      else
        printf(mask[i][j] == 0 ? "   |" : " x |");
    putchar('\n');
  }
}
static inline void print_target() {
  puts("[Target]");
  for (int i = 0; i < nrow; i++) {
    putchar('|');
    for (int j = 0; j < ncol; j++)
      if (target[i][j] == -1)
        printf(" x |");
      else
        printf(" %d |", target[i][j]);
    putchar('\n');
  }
}
static inline void print_piece(piece_t* p) {
  for (int i = 0; i < p->row; i++) {
    for (int j = 0; j < p->col; j++)
      putchar(p_get(p, i, j) ? 'x' : ' ');
    putchar('\n');
  }
}
static inline void print_pieces() {
  puts("[Pieces & Variations]");
  for (int k = 0; k < npieces; k++) {
    int cnt = 0;
    piece_t *p = pieces[k];
    while (p) {
      cnt++;
      print_piece(p); puts("-----");
      p = p->next;
    }
    printf("<< piece %d has %d variations\n", k, cnt);
  }
}

void dfs(int k) {
  if (k == npieces) {
    time_t now = time(NULL);
    printf(">> new solution in %lld sec\n", now - ts_last);
    ts_last = now;
    print_solution();
    return;
  }

  piece_t* p = pieces[k];   // try all variations
  while (p) {
    for (int i = 0; i <= nrow - p->row; i++)      // at all location
      for (int j = 0; j <= ncol - p->col; j++)
        if (p_can_put(p, i, j)) {
          p_put(p, i, j, k + 1);                  // set values=k+1 in mask[][]
          dfs(k + 1);
          p_unput(p, i, j);
        }
    p = p->next;            // next variation
  }
}

static inline void help() {
  printf("Usage: \n");
  printf("   %s\n", bin);
  printf("   %s <month> <date> <weekday>\n", bin);
  exit(-1);
}

// main entry
int main(int argc, char* argv[])  {
  // parse args
  bin = argv[0];
  if (argc == 1) {          // today
    struct tm ts;
    time_t now = time(NULL);
    localtime_s(&ts, &now);
    m = ts.tm_mon + 1;              // 0~11 => 1~12
    d = ts.tm_mday;                 // 1~31
    w = (ts.tm_wday + 6) % 7 + 1;   // 0~6 (Sun. first) => 1~7 (Mon. first)
  } else if (argc == 4) {   // certain day
    m = atoi(argv[1]);
    d = atoi(argv[2]);
    w = atoi(argv[3]);
  } else help();

  // check target
  if (!(1 <= m && m <= 12 && 1 <= d && d <= 31 && 1 <= w && w <= 7)) {
    puts("[Error] invalid target!");
    exit(-2);
  }

  // make piece variations
  for (int k = 0; k < npieces; k++) {
    pieces[k] = &proto[k];   // derive all variations from the prototype
    piece_t *p = p_copy(pieces[k]);   // make a tmp copy

    // rotate
    for (int d = 0; d < 4; d++) {
      p = p_rotate(p);
      p_collect(p, k);
    }
    
    // mirror + rotate
    p = p_mirror(p);
    for (int d = 0; d < 4; d++) {
      p = p_rotate(p);
      p_collect(p, k);
    }

    free(p);
  }

  // init target & current solution
  for (int i = 0; i < nrow; i++)
    for (int j = 0; j < ncol; j++) {
      mask[i][j] = board_value[i][j] ? 0 : -1;    // invalid blocks marked -1, othewise left 0
      switch (board_type[i][j]) {
        case M: target[i][j] = m != board_value[i][j]; break;
        case D: target[i][j] = d != board_value[i][j]; break;
        case W: target[i][j] = w != board_value[i][j]; break;
        case _: target[i][j] = -1; break;
      }
    }
  
  if (debug) { print_pieces(); putchar('\n'); }
  print_board (); putchar('\n');
  print_target(); putchar('\n');

  // rush B!!
  ts_start = ts_last = time(NULL);
  printf("[Solver] search solutions for %s%s %s\n", MN[m], DN[d], WN[w]);
  dfs(0);

  printf(">> done in %lld sec\n", time(NULL) - ts_start);
}

File: solver.js
// Author: Armit 
// Create Time: 2022/08/27
// Update Time: 2022/09/26 

// Usage:
//   node.exe solver.js
//   node.exe solver.js 11 45 14

/* utils */
function StringBuilder() { this._stringArray = new Array(); }
StringBuilder.prototype.reset = function() { this._stringArray.length = 0; }
StringBuilder.prototype.append = function(str) { this._stringArray.push(str); }
StringBuilder.prototype.toString = function(sep='') { return this._stringArray.join(sep); }
const sb = new StringBuilder();

function Array2d() { }
Array2d.prototype.new = function(h, w) {
  return Array.from(Array(h), () => new Array(w));
}
Array2d.prototype.equal = function(A, B) {
  let h = A.length, hh = B.length;
  let w = A[0].length, ww = B[0].length;
  if (h != hh || w != ww) return false;

  for (let i = 0; i < h; i++)
    for (let j = 0; j < w; j++)
      if (A[i][j] != B[i][j])
        return false;
  return true;
}
Array2d.prototype.rotate = function(A) {
  let h = A.length, w = A[0].length;
  let B = Array2d.prototype.new(w, h);
  for (let i = 0; i < w; i++)
    for (let j = 0; j < h; j++)
      B[i][j] = A[j][w - i - 1];
  return B;
}
Array2d.prototype.mirror = function(A) {
  let h = A.length, w = A[0].length;
  let B = Array2d.prototype.new(h, w);
  for (let i = 0; i < h; i++)
    for (let j = 0; j < w; j++)
      B[i][j] = A[i][w - j - 1];
  return B;
}
const Grid = new Array2d();

/* perfcount */
var ts_start;      // ts from dfs() start
var ts_last;       // ts from last solution found

/* readable name mapping */
const MN = [ '', 
  '一月', '二月', '三月', '四月', '五月', '六月',
  '七月', '八月', '九月', '十月', '十一月', '十二月',
];
const DN = [ '', 
  '1日', '2日', '3日', '4日', '5日', '6日', '7日', '8日', '9日', '10日',
  '11日', '12日', '13日', '14日', '15日', '16日', '17日', '18日', '19日', '20日',
  '21日', '22日', '23日', '24日', '25日', '26日', '27日', '28日', '29日', '30日',
  '31日',
];
const WN = [ '', 
  '星期一', '星期二', '星期三', '星期四', '星期五', '星期六', '星期日',
];

/* board definition */
const M = 1, D = 2, W = 3, _ = 0;
const board_type = [
  [ M, M, M, M, M, M, _ ],
  [ M, M, M, M, M, M, _ ],
  [ D, D, D, D, D, D, D ],
  [ D, D, D, D, D, D, D ],
  [ D, D, D, D, D, D, D ],
  [ D, D, D, D, D, D, D ],
  [ D, D, D, W, W, W, W ],
  [ _, _, _, _, W, W, W ],
];
const board_value = [
  [  1,  2,  3,  4,  5,  6,  0 ],
  [  7,  8,  9, 10, 11, 12,  0 ],
  [  1,  2,  3,  4,  5,  6,  7 ],
  [  8,  9, 10, 11, 12, 13, 14 ],
  [ 15, 16, 17, 18, 19, 20, 21 ],
  [ 22, 23, 24, 25, 26, 27, 28 ],
  [ 29, 30, 31,  7,  1,  2,  3 ],
  [  0,  0,  0,  0,  4,  5,  6 ],
];
const pieces = [
  [
    [1, 1, 1, 1],     // ####
  ],
  [
    [1, 1, 1, 1],     // ####
    [1, 0, 0, 0],     // #
  ],
  [
    [0, 1, 1, 1],     //  ###
    [1, 1, 0, 0],     // ##
  ],
  [
    [1, 1, 1],        // ###
    [1, 0, 0],        // #
  ],
  [
    [0, 1, 1],        //  ##
    [1, 1, 0],        // ##
  ],
  [
    [1, 1, 1],        // ###
    [1, 0, 1],        // # #
  ],
  [
    [1, 1, 1],        // ###
    [1, 1, 0],        // ##
  ],
  [
    [1, 1, 1],        // ###
    [1, 0, 0],        // #
    [1, 0, 0],        // #
  ],
  [
    [1, 1, 1],        // ###
    [0, 1, 0],        //  #
    [0, 1, 0],        //  #
  ],
  [
    [0, 1, 1],        //  ##
    [0, 1, 0],        //  #
    [1, 1, 0],        // ##
  ]
].sort(() => .5 - Math.random());   // randomize!
function init_variations() {
  function check_in(newpiece, variations) {
    for (let i in variations)
      if (Grid.equal(newpiece, variations[i]))
        return true;
    return false;
  }

  let variations = Grid.new(pieces.length, 0);
  for (let i = 0; i < pieces.length; i++) {
    let piece = pieces[i];
    variations[i].push(piece);

    // rotate
    let newpiece = piece;
    for (let k = 0; k < 4; k++) {
      newpiece = Grid.rotate(newpiece);
      if (!check_in(newpiece, variations[i]))
        variations[i].push(newpiece);
    }

    // mirror & rotate
    newpiece = Grid.mirror(piece);
    for (let k = 0; k < 4; k++) {
      newpiece = Grid.rotate(newpiece);
      if (!check_in(newpiece, variations[i]))
        variations[i].push(newpiece);
    }
  }
  return variations;
}
const variations = init_variations()

/* puzzle solver */
function make_target_and_mask(m, d, w) {
  let nrow = board_value.length, ncol = board_value[0].length;
  let target = Grid.new(nrow, ncol), mask = Grid.new(nrow, ncol);

  for (let i = 0; i < nrow; i++)
    for (let j = 0; j < ncol; j++) {
      mask[i][j] = board_value[i][j] ? 0 : -1;
      switch (board_type[i][j]) {
        case M: target[i][j] = m != board_value[i][j]; break;
        case D: target[i][j] = d != board_value[i][j]; break;
        case W: target[i][j] = w != board_value[i][j]; break;
        case _: target[i][j] = -1; break;
      }
    }

  return {target: target, mask: mask};
}
function print_grid(grid, name='') {
  if (name) console.log('[' + name + ']');

  let h = grid.length, w = grid[0].length;
  sb.reset();
  for (let i = 0; i < h; i++) {
    sb.append('|');
    for (let j = 0; j < w; j++) {
      if (grid[i][j] == -1) s = 'x';
      else if (grid[i][j] == 0) s = ' ';
      else s = Number(grid[i][j]) - 1;
      sb.append(' ' + s + ' |');
    }
    sb.append('\n');
  }
  console.log(sb.toString());
}

function* dfs(mask, target, v, by_step=false) {
  if (v == variations.length) {
    now = Date.now();
    console.log('>> new solution in ' + Number((now - ts_last) / 1000) + ' sec');
    ts_last = now;
    if (by_step) yield { board: mask, is_solution: true };
    else yield mask;
  }

  let h = mask.length, w = mask[0].length;

  // 检查：形成了单元孤岛
  for (let i = 0; i < h; i++)
    for (let j = 0; j < w; j++)
      if (mask[i][j] == 0) {    // 当前块是空地
        if (target[i][j] == 0) continue;  // 如果目标也是空地那就满足条件，不管

        let N = i - 1  < 0 || mask[i - 1][j] > 0;
        let S = i + 1 >= h || mask[i + 1][j] > 0;
        let W = j - 1  < 0 || mask[i][j - 1] > 0;
        let E = j + 1 >= w || mask[i][j + 1] > 0;
        if (N && S && W && E) return;     // 上下左右都已经填充
      }

  // 对于一块拼图的各种变形方式
  for (let k in variations[v]) {
    let piece = variations[v][k];   
    let hh = piece.length, ww = piece[0].length;

    // 尝试把它放在棋盘的每个地方
    for (let r = 0; r <= h - hh; r++)
      for (let c = 0; c <= w - ww; c++) {
        // 检查：重叠冲突 & 目标被遮盖
        let chk = true;
        for (let rr = r; rr < r + hh && chk; rr++)
          for (let cc = c; cc < c + ww && chk; cc++)
            if (piece[rr-r][cc-c] == 1 && (mask[rr][cc] != 0 || target[rr][cc] == 0))
              chk = false;
        if (!chk) continue;
        
        // 拼入
        let code = 1 + Number(v);
        for (let rr = r; rr < r + hh; rr++)
          for (let cc = c; cc < c + ww; cc++)
            if (piece[rr-r][cc-c])
              mask[rr][cc] = code;    // 拼图块编号1~10
        if (by_step) yield { board: mask, is_solution: false };

        // 递归
        yield* dfs(mask, target, v + 1, by_step);

        // 还原
        for (let rr = r; rr < r + hh; rr++)
          for (let cc = c; cc < c + ww; cc++)
            if (piece[rr-r][cc-c])
              mask[rr][cc] = 0;
        if (by_step) yield { board: mask, is_solution: false };
      }
  }
}
function* solve(month, day, week, by_step=false) {
  let R = make_target_and_mask(month, day, week);
  print_grid(R.target, 'target');
  
  ts_start = ts_last = Date.now();
  yield* dfs(R.mask, R.target, 0, by_step);
}

/* main entry */
if (typeof window === 'undefined') {    // only run in CUI
  // parse args
  argc = process.argv.length;
  if (argc == 2) {          // today
    let now = new Date();
    m = now.getMonth() + 1;
    d = now.getDate();
    w = (now.getDay() + 6) % 7 + 1;
  } else if (argc == 5) {   // certain day
    m = Number(process.argv[2]);
    d = Number(process.argv[3]);
    w = Number(process.argv[4]);
  } else {
    console.log('Usage: ');
    console.log('   node.exe solver.js');
    console.log('   node.exe solver.js <month> <date> <weekday>');
    process.exit(-1);
  }

  // check target
  if (!(1 <= m && m <= 12 && 1 <= d && d <= 31 && 1 <= w && w <= 7)) {
    console.log('[Error] invalid target!');
    process.exit(-2);
  }

  // rush B!!
  console.log('[Solver] search solutions for: ' + MN[m] + DN[d] + ' ' + WN[w]);
  for (let solut of solve(m, d, w))     // SYNTAX: `for (it of gen)`
    print_grid(solut, 'solution');
  
  console.log('>> done in ' + Number((Date.now() - ts_start) / 10) + ' sec');
}